### PR TITLE
Fix network deadlock and improve tests

### PR DESCRIPTION
**Pull request by nearmax**

### CODE REVIEW

File: node/network/src/peer.rs [hunk 1]
```diff
@@ -17,6 +17,7 @@ use tokio::net::TcpStream;
 use tokio::prelude::stream::SplitStream;
 use tokio::timer::Delay;
 use tokio_serde_cbor::Codec;
+use std::sync::RwLockWriteGuard;
```
Comment (bowenwang1996) [inline node/network/src/peer.rs]: parking_lot's `RwLockWriteGuard` is probably faster.

File: node/network/src/peer_manager.rs [hunk 2]
```diff
@@ -163,204 +164,244 @@ impl PeerManager {
             })
             .collect()
     }
+
+    #[cfg(test)]
```
Comment (bowenwang1996) [inline node/network/src/peer_manager.rs]: Why is this not in the tests module?

Comment (nearmax) [reply node/network/src/peer_manager.rs]: Because it takes `self`.

Comment (bowenwang1996) [reply node/network/src/peer_manager.rs]: You can do a `impl` in the tests module

File: node/network/src/peer.rs (hunk 1 above)
Comment (nearmax) [reply node/network/src/peer.rs]: It does not handle lock poisoning, which is not safe: https://www.reddit.com/r/rust/comments/4kng0h/parking_lot_highly_optimized_synchronization/d3gh3gt

File: node/network/src/peer_manager.rs (hunk 2 above)
Comment (nearmax) [reply node/network/src/peer_manager.rs]: Good point.

Comment (nearmax) [reply node/network/src/peer_manager.rs]: Done.

**Pull request merged**